packages/flutter/lib/src/material/mergeable_material.dart

```diff
@@ -0,0 +1,562 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+import 'package:flutter/rendering.dart';
+
+import 'dart:ui' show lerpDouble;
+
+/// A class that abstracts over a Material slice and a Material gap.
+///
+/// All [MergeableMaterialItem] objects need a [LocalKey].
+abstract class MergeableMaterialItem {
+  MergeableMaterialItem(this.key) {
+    assert(key != null);
+  }
+
+  final LocalKey key;
+}
+
+/// A class that can be used as a child to [MergeableMaterial]. It is a slice
+/// of [Material] that animates merging with other slices.
+///
+/// All [MaterialSlice] objects need a [LocalKey].
+class MaterialSlice extends MergeableMaterialItem {
+  /// Creates a slice of [Material] that's mergeable within a
+  /// [MergeableMaterial].
+  MaterialSlice({
+    LocalKey key,
+    this.child
+  }) : super(key);
+
+  /// The widget below this widget in the tree.
+  final Widget child;
+
+  @override
+  String toString() {
+    return 'MergeableSlice(key: $key, child: $child)';
+  }
+}
+
+/// A class that represents a gap within [MergeableMaterial].
+///
+/// All [MaterialGap] objects need a [LocalKey].
+class MaterialGap extends MergeableMaterialItem {
+  /// Creates a Material gap with a given size.
+  MaterialGap({
+    LocalKey key,
+    this.size: 16.0
+  }) : super(key);
+
+  /// The size of the gap, irrespective of the axis used.
```

Comment:
I'd phrase this as "The main axis extent of this gap. For example, if the [MergableMaterial] is vertical, then this is the height of the gap."